dissect/target/plugins/os/windows/catroot.py

```diff
@@ -60,67 +114,134 @@ def catroot(self):
         """
-        # So asn1crypt dies when parsing these files, so we kinda bruteforce it
-        # Look for the object identifiers of various hash types, and parse from there
-        # We don't do any further checking, just traverse according to a known structure
-        # If an exception occurs, we're not looking at the right structure.
-        for d in self.catrootdir.iterdir():
-            if not d.is_dir():
+        # As far as known, Microsoft uses its own implementation to store the digest in the
+        # encap_content_info along with an optional file hint. Here we parse the digest values
+        # ourselves by looking for the corresponding digest needles in the raw encap_content_info
+        # data. Furthermore, we try to find the file hint if it is present in that same raw data.
+        catroot_info = self.catroot_dirs[CatrootFileType.CATROOT_FILE]
+        for file in catroot_info.base_path.rglob(catroot_info.pattern):
+            if not file.is_file():
                 continue
 
-            for f in d.iterdir():
-                buf = f.open().read()
+            try:
+                with file.open("rb") as fh:
+                    buf = fh.read()
+
+                # TODO: Parse other data in the content info
+                content_info = ContentInfo.load(buf)["content"]
+
+                digest_type = content_info["digest_algorithms"].native[0].get("algorithm")
+                encap_contents = content_info["encap_content_info"].contents
+                needle = NEEDLES[digest_type]
+
+                digests = []
+                for offset in findall(encap_contents, needle):
+                    # 4 bytes before the digest type, a sequence starts
+                    objseq = Sequence.load(encap_contents[offset - 4 :])
+                    # The second entry in the sequence is the digest string
+                    raw_digest = objseq[1].native
+                    hexdigest = raw_digest.hex()
+
+                    file_digest = digest()
+                    if len(hexdigest) == 32:
+                        file_digest.md5 = hexdigest
+                    elif len(hexdigest) == 40:
+                        file_digest.sha1 = hexdigest
+                    elif len(hexdigest) == 64:
+                        file_digest.sha256 = hexdigest
+
+                    digests.append(file_digest)
+
+                # Finding the hint in encap_content_info is on best effort. In most of the cases,
+                # there is a key "PackageName" available. We first try to parse the corresponding
+                # value if it is present. If this does not succeed, we might be dealing with catroot
+                # files containing the "hint" needle.
+                # If both methods do not result in a file hint, there is either no hint available or
+                # the format is not yet known and therefore not supported.
+                hints = []
+                try:
+                    if offset:
```

Comment:
I still get an error, because, if there are no offsets, this is not initialized.

`An error occurred while parsing the hint for catroot file sysvol/windows/system32/catroot/.../OEMBIOS.CAT: local variable 'offset' referenced before assignment`